spec/models/user_spec.rb

```diff
@@ -0,0 +1,171 @@
+    it{should have_many(:following_books)}
+    it{should have_many(:following_authors)}
+    it{should have_many(:following_users)}
+    it{should have_many(:comment_books)}
+    it{should have_many(:comment_authors)}
+    it{should have_many(:like_books)}
+    it{should have_many(:like_authors)}
+    it{should have_many(:passive_follows).dependent(:destroy)}
+    it{should have_many(:followers)}
+  end
+
+  describe "Validates" do
+    subject{FactoryBot.build(:user)}
+    context "with name field " do
+      it{should validate_presence_of(:name)}
+      it{should validate_length_of(:name)
+        .is_at_most(Settings.length.user_name_max)}
+    end
+
+    context "with email field" do
+      it{should validate_presence_of(:email)}
+      it{
+        should validate_length_of(:email)
+          .is_at_most(Settings.length.email_max)}
+      it{should validate_uniqueness_of(:email).case_insensitive}
+      it{should allow_value("a@b.c").for(:email)}
+      it{should_not allow_value("s").for(:email)}
+    end
+
+    context "with address field" do
+      it{should validate_presence_of(:address)}
+      it{should validate_length_of(:address).is_at_least(Settings.length.password_min)}
+    end
+
+    context "with phone field" do
+      it{should validate_presence_of(:phone)}
+      it{should allow_value("0393203261").for(:phone)}
+      it{should_not allow_value("020220").for(:phone)}
+    end
+
+    context "with password field" do
+      it{should validate_length_of(:password).is_at_least(Settings.length.password_min)}
+      it{should validate_confirmation_of(:password)}
+    end
+  end
+
+  describe "#is_admin?" do
+    let(:user){FactoryBot.create :user}
+    it "return true when user is admin" do
+      user.is_admin = true
+      expect(user.is_admin?).to eq true
+    end
+    it "return false when user is not admin" do
+      expect(user.is_admin?).to eq false
+    end
+  end
+
+  describe ".activate" do
+    let(:user){FactoryBot.create :user}
+    it "activated should equal to true" do
+      expect{user.activate} .to change{user.activated}
+    end
+  end
+
+  describe "#authenticated?" do
+    let(:user){FactoryBot.create :user}
+    it "should returns true when correct passowrd" do
+      expect(user.authenticated?(:password, "password!@#")).to eq true
+    end
+
+    it "should returns false when incorrect passowrd" do
+      expect(user.authenticated?(:password, "password!@1#")).to eq false
+    end
+
+    it "should returns false when user digest = nil" do
+      expect(user.authenticated?(:activation, "password!@1#")).to eq false
+    end
+  end
+
+  describe ".send_password_reset_email" do
+    let(:user){FactoryBot.create :user}
+    it "should sends reset email" do
+      user.create_reset_digest
+      expect (user.send_password_reset_email).to change {ActionMailer::Base.deliveries.count}.by(1)
+    end
+  end
+
+  describe ".send_activation_email" do
+    let(:user){FactoryBot.create :user}
```

Comment:
phân biệt let và let!

Reply:
- let sẽ được gọi mỗi lần khi một khối it bắt đầu chạy
 - let! sẽ được gọi trước khi tất cả các khối it bắt đầu chạy và chỉ chạy một lần duy nhất ở lần đầu tiên đó